pkg/proxy/userspace/proxysocket.go

```diff
@@ -133,15 +133,19 @@ func (tcp *tcpProxySocket) ProxyLoop(service proxy.ServicePortName, myInfo *serv
 			continue
 		}
 		glog.V(2).Infof("Accepted TCP connection from %v to %v", inConn.RemoteAddr(), inConn.LocalAddr())
-		outConn, err := tryConnect(service, inConn.(*net.TCPConn).RemoteAddr(), "tcp", proxier)
-		if err != nil {
-			glog.Errorf("Failed to connect to balancer: %v", err)
-			inConn.Close()
-			continue
-		}
-		// Spin up an async copy loop.
-		go proxyTCP(inConn.(*net.TCPConn), outConn.(*net.TCPConn))
+		go connectToBackend(service, inConn, proxier)
```

Comment:
I'm sorry I didn't see this earlier - why is this done in the TCP-specific path?  The TODO that I think you're referring to is actually in the generic (TCP and UDP) path - UDP matters too. (also you should remove the TODO when you fix it :)

What about something like:

* rename `tryConnect` to `connectAndProxy`.
* pass a func(in, out net.Conn) to connectAndProxy() - one for TCP, one for UDP in their respective paths
* let connectAndProxy() call through to the function, which calls through to the proxy loop
* run connectAndProxy() as a goroutine

UDP is a little tricker, but we should solve both,